### PR TITLE
Refactor the questionnaire response validator

### LINKED ISSUE #286: Validate data type and item nesting within answers for repeated answers while prefilling UI with a Questionnaire response
We need to validate the Questionnaire Response structure with that of the Questionnaire before prefilling the UI. This is to make sure that this particular questionnaire response has been generated by answering the current questionnaire.
Currently we only validate nesting of item within group and the linkId.

We will also have to add checks to make sure:

1. The data type of the questionnaire-item and questionnaireResponse-item match
2. Item nesting within answers #274 

This is an improvement related to #244

### PR DESCRIPTION
**Pull request by jingtang10**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #286

**Description**
Refactor the questionnaire response validator. Broken the `checkQuestionnaireResponse` function into multiple sub-functions for better structure and readability.

**Alternative(s) considered**
NA

**Type**
Code health

**Screenshots (if applicable)**
NA

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/validation/QuestionnaireResponseValidator.kt [hunk 1]
```diff
@@ -69,88 +70,166 @@ object QuestionnaireResponseValidator {
...
-        )
+    while (questionnaireResponseInputItemIterator.hasNext()) {
+      val questionnaireResponseItem = questionnaireResponseInputItemIterator.next()
+      var questionnaireItem: Questionnaire.QuestionnaireItemComponent?
+      do {
+        require(questionnaireItemIterator.hasNext()) {
+          "Missing questionnaire item for questionnaire response item ${questionnaireResponseItem.linkId}"
+        }
+        questionnaireItem = questionnaireItemIterator.next()
+      } while (questionnaireItem!!.linkId != questionnaireResponseItem.linkId)
```
Comment (aditya-07) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/validation/QuestionnaireResponseValidator.kt:126]: Can we have a `map of linkId & questionnaireItem` and do a lookup by `questionnaireResponseItem.linkId` to find corresponding  `questionnaireItem`?

File: datacapture/src/test/java/com/google/android/fhir/datacapture/validation/QuestionnaireResponseValidatorTest.kt [hunk 2]
```diff
@@ -159,19 +171,1072 @@ class QuestionnaireResponseValidatorTest {
...
+
+  // @Test
+  // fun `check fails if questionnaire response does not match questionnaire`() {
+  //   assertCheckQuestionnaireResponseThrowsIllegalArgumentException(
+  //     Questionnaire().apply { url = "questionnaire-1" },
+  //     QuestionnaireResponse().apply { questionnaire = "questionnaire-2" },
+  //     "Mismatching Questionnaire questionnaire-1 and QuestionnaireResponse (for Questionnaire
+  // questionnaire-2)"
+  //   )
+  // }
```
Comment (aditya-07) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/validation/QuestionnaireResponseValidatorTest.kt:198]: Remove it.

**Review by jingtang10 — COMMENTED** [review]
thanks @aditya-07 - see comments below

**Review by aditya-07 — APPROVED** [review]
LGTM!

**Pull request merged**